sdk/eventhubs/azure-messaging-eventhubs-checkpointstore-blob/src/main/java/com/azure/messaging/eventhubs/checkpointstore/blob/MetricsHelper.java

```diff
@@ -0,0 +1,188 @@
+        if (!isEnabled || !(lastSequenceNumber.isEnabled() && checkpointCounter.isEnabled())) {
+            return;
+        }
+
+        if (!maxCapacityReached && (seqNoSubscriptions.size() >= MAX_ATTRIBUTES_SETS || common.size() >= MAX_ATTRIBUTES_SETS)) {
+            LOGGER.error("Too many attribute combinations are reported for checkpoint metrics, ignoring any new dimensions.");
+            maxCapacityReached = true;
+        }
+
+        if (lastSequenceNumber.isEnabled() && success) {
+            updateCurrentValue(attributesId, checkpoint);
+        }
+
+        if (checkpointCounter.isEnabled()) {
+            TelemetryAttributes attributes = null;
+            if (success) {
+                attributes = getOrCreate(checkpointSuccess, attributesId, checkpoint, "ok");
+            } else {
+                attributes = getOrCreate(checkpointFailure, attributesId, checkpoint, "error");
+            }
+            if (attributes != null) {
+                checkpointCounter.add(1, attributes, Context.NONE);
+            }
+        }
+    }
+
+    private TelemetryAttributes getOrCreate(ConcurrentHashMap<String, TelemetryAttributes> source, String attributesId, Checkpoint checkpoint, String status) {
+        if (maxCapacityReached) {
+            return source.get(attributesId);
+        }
+
+        return source.computeIfAbsent(attributesId, i -> meter.createAttributes(createAttributes(checkpoint, status)));
+    }
+
+    private Map<String, Object> createAttributes(Checkpoint checkpoint, String status) {
+        Map<String, Object> attributesMap = new HashMap<>();
+        attributesMap.put(HOSTNAME_KEY, checkpoint.getFullyQualifiedNamespace());
+        attributesMap.put(ENTITY_NAME_KEY, checkpoint.getEventHubName());
+        attributesMap.put(PARTITION_ID_KEY, checkpoint.getPartitionId());
+        attributesMap.put(CONSUMER_GROUP_KEY, checkpoint.getConsumerGroup());
+        if (status != null) {
+            attributesMap.put(STATUS_KEY, status);
+        }
+
+        return attributesMap;
+    }
+
+    private void updateCurrentValue(String attributesId, Checkpoint checkpoint) {
+        if (checkpoint.getSequenceNumber() == null) {
+            return;
+        }
+
+        CurrentValue valueSupplier;
+        if (maxCapacityReached) {
+            valueSupplier = seqNoSubscriptions.get(attributesId);
+        } else {
+
+            TelemetryAttributes attributes = getOrCreate(common, attributesId, checkpoint, null);
+            if (attributes == null) {
+                return;
+            }
+
+            valueSupplier = seqNoSubscriptions.computeIfAbsent(attributesId, a -> {
+                AtomicReference<Long> lastSeqNo = new AtomicReference<>();
+                return new CurrentValue(lastSequenceNumber.registerCallback(() -> lastSeqNo.get(), attributes), lastSeqNo);
+            });
+        }
+
+        if (valueSupplier != null) {
```

Comment:
nit; not sure but if we declare` valueSupplier` as `final`  does that help removing this `if` check?

Reply:
it could be that it's null if there are too many attributes, but I tried to make it a little bit more readable :)